src/main/java/org/veupathdb/service/eda/ss/model/reducer/StreamIntersectMerger.java

```diff
@@ -1,92 +1,131 @@
 
 import java.util.*;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
 
 
 /**
- * This class is used to take the intersection of two or more streams of sequenced subsets.
+ * This class is used to take the intersection of two or more streams of sequenced longs. The class produces a
+ * stream of output longs, each of which are present in ALL input streams. This behavior is only well defined if
+ * all input streams are sorted.
  */
 public class StreamIntersectMerger implements Iterator<Long> {
-  private final PeekableIterator[] streams;
-  private final int[] nextStream;
-  private Long currentElement;
-  private int currentStreamIdx;
-  private PeekableIterator currentStream;
+  private final PeekableIterator[] peekableIdIndexStreams;
+  private Long nextOutputIndex;
+  private PeekableIterator currentIdIndexStream;
   private boolean hasStarted;
+  private RingLinkedList streamRing;
 
   /**
    * @param sortedStreams Collection of sorted streams to merge by intersection.
    */
   public StreamIntersectMerger(List<Iterator<Long>> sortedStreams) {
-    this.streams = sortedStreams.stream()
+    // If no input streams are provided, it should act as an "empty" Iterator.
+    // If any input stream is "empty", should act as an "empty" Iterator since we are intersecting Iterators.
+    if (sortedStreams.isEmpty() || !sortedStreams.stream().allMatch(Iterator::hasNext)) {
+      hasStarted = true;
+      peekableIdIndexStreams = null;
+      nextOutputIndex = null;
+      return;
+    }
+    // Convert iterators into peekable iterators.
+    this.peekableIdIndexStreams = sortedStreams.stream()
         .map(PeekableIterator::new)
         .toArray(PeekableIterator[]::new);
-    this.currentStream = Arrays.stream(streams)
-        .max(Comparator.comparing(iter -> iter.peek()))
-        .get();
-    this.currentStreamIdx = IntStream.range(0, streams.length)
-        .filter(i -> streams[i] == currentStream)
-        .findFirst()
-        .getAsInt();
-    // Cache a pointer to the nextStream to avoid excessive modulo operations.
-    this.nextStream = IntStream.range(0, streams.length)
-        .map(i -> (i + 1) % streams.length)
-        .toArray();
+    this.streamRing = new RingLinkedList(peekableIdIndexStreams);
+    this.currentIdIndexStream = streamRing.cursor.currentStream;
     hasStarted = false;
   }
 
   @Override
   public boolean hasNext() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    if (currentElement == null) {
-      return false;
-    } else {
-      return true;
-    }
+    return nextOutputIndex != null;
   }
 
   @Override
   public Long next() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    Long next = currentElement;
-    currentStream.next();
-    setCurrentElement();
+    Long next = nextOutputIndex;
+    currentIdIndexStream.next();
```

Comment:
Maybe put some code here to fulfill the Iterator contract of NoSuchElementException vs returning null, e.g.
```
    if (next == null) {
      throw new NoSuchElementException("No more elements in the iterator");
    }
```

Reply:
Fixed